src/modules/ExactTracking.jl

```diff
@@ -38,19 +39,128 @@ end
+  @assert isnothing(winv) || (size(winv,1) == 3 && size(winv,2) == 3) "The inverse rotation matrix must be either `nothing` or 3x3 for patch!. Received $winv"
+  @inbounds begin @FastGTPSA! begin
+    # Temporary momentum [δp, pz]
+    work[i,1] = 1 + v[i,PZI]                                 # rel_p
+    work[i,2] = sqrt(work[i,1]^2 - v[i,PXI]^2 - v[i,PYI]^2)  # ps_0
+  end end
+    # Only apply rotations if needed
+    if isnothing(winv)
+      @inbounds begin @FastGTPSA! begin
+      # No rotation case
+      v[i,XI] -= dx
+      v[i,YI] -= dy
+      
+      # Apply t_offset
+      v[i,ZI] += work[i,1]*sqrt(work[i,1]/(work[i,1]^2+tilde_m^2))*C_LIGHT*dt
+
+      # Drift to face
+      v = exact_drift!(i, v, work, -dz, tilde_m, gamsqr_0, beta_0)
+      end end
+    else
+      @inbounds begin @FastGTPSA! begin
+      # Translate position vector [x, y]
+      work[i,3] = v[i,XI] - dx                                # x_0
+      work[i,4] = v[i,YI] - dy                                # y_0
+
+      # Temporary momentum vector [px, py]
+      work[i,5] = v[i,PXI]                                    # px_0
+      work[i,6] = v[i,PYI]                                    # py_0
+      
+      # Transform position vector [x - dx, y - dy, -dz]
+      v[i,XI]   = winv[1,1]*work[i,3] + winv[1,2]*work[i,4] - winv[1,3]*dz
+      v[i,YI]   = winv[2,1]*work[i,3] + winv[2,2]*work[i,4] - winv[2,3]*dz
+      work[i,7] = winv[3,1]*work[i,3] + winv[3,2]*work[i,4] - winv[3,3]*dz  # s_f
+      
+      # Transform momentum vector [px, py, ps]
+      v[i,PXI]  = winv[1,1]*work[i,5] + winv[1,2]*work[i,6] + winv[1,3]*work[i,2]
+      v[i,PYI]  = winv[2,1]*work[i,5] + winv[2,2]*work[i,6] + winv[2,3]*work[i,2]
+      work[i,8] = winv[3,1]*work[i,5] + winv[3,2]*work[i,6] + winv[3,3]*work[i,2] # ps_f
+
+      # Apply t_offset
+      v[i,ZI] += work[i,1]/sqrt(work[i,1]^2+tilde_m^2)*C_LIGHT*dt
+
+      # Drift to face
+      v[i,XI] -= work[i,7] * v[i,PXI] / work[i,8]
+      v[i,YI] -= work[i,7] * v[i,PYI] / work[i,8]
+      v[i,ZI] += work[i,7] * work[i,1] / work[i,8] + L*work[i,1]*sqrt((1+tilde_m^2)/(work[i,1]^2+tilde_m^2))
+      end end
+    end
+  return v
+end
+
+
+# Utility functions ============================================================
+
+# Rotation matrix
+"""
+  w_matrix(x_rot, y_rot, z_rot)
+
+Constructs a rotation matrix based on the given Bryan-Tait angles.
+
+Bmad/SciBmad follows the MAD convention of applying z, x, y rotations in that order.
+Furthermore, in ReferenceFrameRotations, the rotation angles around Y and Z axes 
+are defined as negative of the SciBmad `y_rot` and `z_rot`.
+
+The inverse matrix reverses the order of operations and their signs.
+
+
+Arguments:
+- `x_rot::Real`: Rotation angle around the x-axis.
+- `y_rot::Real`: Rotation angle around the y-axis.
+- `z_rot::Real`: Rotation angle around the z-axis.
```

Comment:
```suggestion
- `x_rot::Number`: Rotation angle around the x-axis.
- `y_rot::Number`: Rotation angle around the y-axis.
- `z_rot::Number`: Rotation angle around the z-axis.
```

according to the [docs](https://juliaspace.github.io/ReferenceFrameRotations.jl/stable/lib/library/#ReferenceFrameRotations.angle_to_rot-Tuple{Number,%20Symbol}) these do not need to be real but `Number`. The `TPS` type for example only subtypes `Real` and not `Number`